<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>

In [1]:
import pandas as pd
import numpy as np
import requests
import json
from pandas.io.json import json_normalize 
# import k-means from clustering stage
from sklearn.cluster import KMeans


In [2]:
!pip install lxml
!pip install geocoder
import geocoder

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim 

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library


Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.



## 1. Data Source and Cleaning 

Reading tables using pandas from Wikipidia page


In [3]:
tables = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header=0,keep_default_na=False)
toronto_data = tables[0]
toronto_data.shape

(180, 3)

Ignore cells with a borough that is Not assigned. 

In [4]:
toronto_data.drop(toronto_data.loc[toronto_data['Borough']=='Not assigned'].index, inplace=True)
#toronto_data = toronto_data[toronto_data.Borough != 'Not assigned'] -- alternate code
toronto_data.shape

(103, 3)

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [5]:
toronto_data['Neighbourhood'] = np.where(toronto_data['Neighbourhood'] == 'Not assigned', toronto_data['Borough'], toronto_data['Neighbourhood']) # It is not needed since NA to data.
toronto_data[toronto_data['Neighbourhood']=='Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

More than one neighborhood can exist in one postal code area. These two rows will be combined into one row with the neighborhoods separated with a comma.

In [6]:
toronto_data.groupby(by=['Postal Code','Borough']).agg({'Neighbourhood':', '.join})
print(toronto_data.head())
toronto_data.shape

  Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government


(103, 3)

### Q1 Result:
### TORONTO DATA SHAPE : (103, 3)

Geographical coordinates of each postal code: 
# geocoder didn't work properly

In [7]:
import geocoder # import geocoder
codes = toronto_data['Postal Code'].tolist()  
codes = codes[:2]
for postal_code in codes:
# initialize your variable to None
    lat_lng_coords = None
    cnt = 20
    print("Code :", postal_code)

# loop until you get the coordinates
    while(lat_lng_coords is None and cnt == 0):
      g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
      lat_lng_coords = g.latlng
      cnt -= 1
    if lat_lng_coords is not None:
        latitude = lat_lng_coords[0]
        longitude = lat_lng_coords[1]
    print(lat_lng_coords)

Code : M3A
None
Code : M4A
None


Geographical coordinates of each postal code using csv file

In [8]:
geo_cord = pd.read_csv('http://cocl.us/Geospatial_data')
geo_cord

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

# Q2

Merging geo cordinates to the toronto data

In [9]:
toronto_data_with_geo = pd.merge(toronto_data,geo_cord,on='Postal Code') #default inner join 
toronto_data_with_geo

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

# Q3
I followed the same approach as New York data and made my analysis using K-means clustering.

<h3> 1. Explore Neighborhoods in Toronto City </h3>

In [16]:
neighborhoods = toronto_data_with_geo[['Borough', 'Neighbourhood', 'Latitude', 'Longitude']]
neighborhoods.head()
#neighborhoods.shape

Borough                                Neighbourhood   Latitude  \
0        North York                                    Parkwoods  43.753259   
1        North York                             Victoria Village  43.725882   
2  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
3        North York             Lawrence Manor, Lawrence Heights  43.718518   
4  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [17]:
neighborhoods.shape

(103, 4)

#### Use geopy library to get the latitude and longitude values of Toronto City.

In [18]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tr_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [19]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [20]:
tor_data = neighborhoods[neighborhoods['Borough'].str.contains('Toronto')].reset_index(drop=True)
tor_data.head()

print(tor_data.shape)
print(tor_data.loc[0, 'Neighbourhood'])

(39, 4)
Regent Park, Harbourfront


#### Define Foursquare API to get json data

In [22]:
neighborhood_latitude = tor_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = tor_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = tor_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


In [23]:
CLIENT_ID = 'CROR0TLNBZEMVIFU11M2MWBAWHHWPCBJ1UCTEPWX5UZ1NH0U' # your Foursquare ID
CLIENT_SECRET = 'TPMSNRTQ054CIWNCXJFL2RWD0M2RLCF5NQXUCSA024FJCRRN' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
# type your answer here

LIMIT = 50 # limit of number of venues returned by Foursquare API



radius = 500 # define radius

 # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

Your credentails:
CLIENT_ID: CROR0TLNBZEMVIFU11M2MWBAWHHWPCBJ1UCTEPWX5UZ1NH0U
CLIENT_SECRET:TPMSNRTQ054CIWNCXJFL2RWD0M2RLCF5NQXUCSA024FJCRRN


'https://api.foursquare.com/v2/venues/explore?&client_id=CROR0TLNBZEMVIFU11M2MWBAWHHWPCBJ1UCTEPWX5UZ1NH0U&client_secret=TPMSNRTQ054CIWNCXJFL2RWD0M2RLCF5NQXUCSA024FJCRRN&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=50'

In [24]:
results = requests.get(url).json()
#results -- data is long

In [25]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [26]:
venues = results['response']['groups'][0]['items']
venues

nearby_venues = json_normalize(venues) # flatten JSON
nearby_venues.head()
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]


nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  after removing the cwd from sys.path.


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4          Impact Kitchen           Restaurant  43.656369 -79.356980

In [27]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

45 venues were returned by Foursquare.


### 2. Explore Neighborhoods in Toronto

In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [29]:
tor_venues = getNearbyVenues(names=tor_data['Neighbourhood'],
                                   latitudes=tor_data['Latitude'],
                                   longitudes=tor_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [30]:
print(tor_venues.shape)
tor_venues.head()



(1190, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [31]:
tor_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            50   
Brockton, Parkdale Village, Exhibition Place                           26   
Business reply mail Processing Centre, South Ce...                     18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     50   
Christie                                                               17   
Church and Wellesley                                                   50   
Commerce Court, Victoria Hotel                                         50   
Davisville                                                             34   
Davisville North                                                       10   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                 50   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                               50   
Harbourfront East, Union Station, Toronto Islands                      50   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              50   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               46   
Moore Park, Summerhill East                                             4   
North Toronto West, Lawrence Park                                      17   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            34   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               50   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     39   
St. James Town                                                         50   
St. James Town, Cabbagetown                                            40   
Stn A PO Boxes                                                         50   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    20   
The Beaches                                                             4   
The Danforth West, Riverdale                                           41   
Toronto Dominion Centre, Design Exchange                               50   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             50   
Brockton, Parkdale Village, Exhibition Place                            26   
Business reply mail Processing Centre, South Ce...                      18   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      50   
Christie                                                                17   
Church and Wellesley                                                    50   
Commerce Court, Victoria Hotel                                          50   
Davisville                                                              34   
Davisville North                    

In [32]:
print('There are {} uniques categories.'.format(len(tor_venues['Venue Category'].unique())))

There are 216 uniques categories.


## 3. Analyze Each Neighborhood

In [34]:
# one hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighborhood'] = tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  ...  Thai Restaurant  Theater  Theme Restaurant  \
0         0  ...                0        0                 0   
1         0  ...                0        0                 0   
2         0  ...                0        0                 0   
3         0  ...                0        0                 0   
4         0  ...                0        0                 0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  
0                 0                      0         0  
1                 0                      0         0  
2                 0                      0         0  
3                 0                      0         0  
4                 0                      0         0  

[5 rows x 216 columns]

In [35]:
tor_onehot.shape

(1190, 216)

In [37]:
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped
tor_grouped.shape

(39, 216)

In [38]:
num_top_venues = 5

for hood in tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = tor_grouped[tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.08
1        Cocktail Bar  0.04
2            Beer Bar  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
                   venue  freq
0                   Café  0.12
1            Coffee Shop  0.08
2         Breakfast Spot  0.08
3  Performing Arts Venue  0.08
4                 Bakery  0.08


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0    Light Rail Station  0.11
1  Gym / Fitness Center  0.06
2         Auto Workshop  0.06
3            Comic Shop  0.06
4           Pizza Place  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2   Harbor / Marina  0.06
3          Boutique  0.06
4  Sculpture Garden  0.06


----Central Bay Stree

In [39]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [40]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0    Seafood Restaurant            Cheese Shop        Farmers Market   
1        Breakfast Spot  Performing Arts Venue           Coffee Shop   
2         Garden Center             Skate Park               Brewery   
3        Airport Lounge          Boat or Ferry              Boutique   
4        Sandwich Place     Italian Restaurant       Bubble Tea Shop   

  5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
0              Beer Bar            Restaurant                      Bakery   
1                Bakery          Climbing Gym               Burrito Place   
2                   Spa         Burrito Place                  Restaurant   
3   Rental Car Location      Sculpture Garden                       Plane   
4          Burger Joint                  Café  Modern European Restaurant   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0                  Café          Cocktail Bar               Liquor Store  
1             Nightclub            Restaurant         Italian Restaurant  
2               Butcher      Recording Studio             Farmers Market  
3       Harbor / Marina           Coffee Shop                        Bar  
4   Japanese Restaurant             Juice Bar  Middle Eastern Restaurant

## 4. Cluster Neighborhoods

In [41]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

tor_grouped_clustering = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 3, 0, 0, 3, 0, 0, 3, 0], dtype=int32)

In [42]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

tor_merged = tor_data
#df.rename(columns={'oldName1': 'newName1', 'oldName2': 'newName2'}, inplace=True)
tor_merged.rename(columns={'Neighbourhood':'Neighborhood'}, inplace=True)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tor_merged.head() # check the last columns!

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4      East Toronto                                  The Beaches  43.676357   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                   Pub   
1 -79.389494               0           Coffee Shop           Yoga Studio   
2 -79.378937               0           Coffee Shop                  Café   
3 -79.375418               0                  Café        Cosmetics Shop   
4 -79.293031               3                 Trail                   Pub   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park                Bakery                  Café   
1     College Cafeteria         Smoothie Shop              Beer Bar   
2      Ramen Restaurant             Bookstore               Theater   
3           Coffee Shop            Restaurant                  Park   
4     Health Food Store              Wine Bar          Cupcake Shop   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0        Breakfast Spot            Restaurant                Theater   
1        Sandwich Place                  Café  Portuguese Restaurant   
2        Clothing Store    Italian Restaurant         Cosmetics Shop   
3                 Hotel             Gastropub     Seafood Restaurant   
4      Doner Restaurant               Dog Run    Distribution Center   

  9th Most Common Venue 10th Most Common Venue  
0         Historic Site            Event Space  
1                  Park     College Auditorium  
2    Seafood Restaurant                    Spa  
3              Creperie         Farmers Market  
4        Discount Store                  Diner

In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

In [45]:
tor_merged.loc[tor_merged['Cluster Labels'] == 0, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Neighborhood  \
0                           Regent Park, Harbourfront   
1         Queen's Park, Ontario Provincial Government   
2                            Garden District, Ryerson   
3                                      St. James Town   
5                                         Berczy Park   
6                                  Central Bay Street   
8                            Richmond, Adelaide, King   
10  Harbourfront East, Union Station, Toronto Islands   
11                           Little Portugal, Trinity   
13           Toronto Dominion Centre, Design Exchange   
16                     Commerce Court, Victoria Hotel   
17                                    Studio District   
20                                   Davisville North   
23                  North Toronto West, Lawrence Park   
30          Kensington Market, Chinatown, Grange Park   
32  CN Tower, King and Spadina, Railway Lands, Har...   
34                                     Stn A PO Boxes   
36             First Canadian Place, Underground city   
37                               Church and Wellesley   

            1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                     Coffee Shop                   Pub                  Park   
1                     Coffee Shop           Yoga Studio     College Cafeteria   
2                     Coffee Shop                  Café      Ramen Restaurant   
3                            Café        Cosmetics Shop           Coffee Shop   
5                     Coffee Shop    Seafood Restaurant           Cheese Shop   
6                     Coffee Shop        Sandwich Place    Italian Restaurant   
8                     Coffee Shop                  Café            Steakhouse   
10                       Aquarium           Coffee Shop                 Plaza   
11                            Bar      Asian Restaurant           Coffee Shop   
13                    Coffee Shop                  Café    Seafood Restaurant   
16                    Coffee Shop                  Café            Restaurant   
17                           Café           Coffee Shop   American Restaurant   
20                           Park                   Gym        Breakfast Spot   
23                 Clothing Store           Coffee Shop           Yoga Studio   
30  Vegetarian / Vegan Restaurant                  Café           Coffee Shop   
32                Airport Service        Airport Lounge         Boat or Ferry   
34                           Café           Coffee Shop            Restaurant   
36                           Café           Coffee Shop            Restaurant   
37               Sushi Restaurant            Restaurant           Coffee Shop   

   4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
0                 Bakery                  Café         Breakfast Spot   
1          Smoothie Shop              Beer Bar         Sandwich Place   
2              Bookstore               Theater         Clothing Store   
3             Restaurant                  Park                  Hotel   
5         Farmers Market              Beer Bar             Restaurant   
6        Bubble Tea Shop          Burger Joint                   Café   
8           Concert Hall   American Restaurant                  Hotel   
10               Brewery                  Café                   Park   
11                  Café           Men's Store             Restaurant   
13            Restaurant                 Hotel               Beer Bar   
16                 Hotel                   Gym    Japanese Restaurant   
17                Bakery               Brewery              Gastropub   
20        Sandwich Place      Department Store           Dance Studio   
23  Fast Food Restaurant    Chinese Restaurant                   Café   
30    Mexican Restaurant          Dessert Shop  Vietnamese Restaurant   
32              Boutique   Rental Car Location       Sculpture Garden   
34          Cocktail Bar        Farmer

In [46]:
tor_merged.loc[tor_merged['Cluster Labels'] == 1, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
21  Forest Hill North & West, Forest Hill Road Park      Sushi Restaurant   
33                                         Rosedale                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                  Park         Jewelry Store                 Trail   
33            Playground                 Trail              Wine Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21              Wine Bar      Doner Restaurant               Dog Run   
33          Cupcake Shop      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21   Distribution Center        Discount Store                  Diner  
33   Distribution Center        Discount Store                  Diner

In [47]:
tor_merged.loc[tor_merged['Cluster Labels'] == 2, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
19     Roselawn                Garden          Home Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19              Wine Bar          Cupcake Shop            Donut Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19      Doner Restaurant               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
19        Discount Store                  Diner

In [48]:
tor_merged.loc[tor_merged['Cluster Labels'] == 3, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
4                                         The Beaches                 Trail   
7                                            Christie         Grocery Store   
9                        Dufferin, Dovercourt Village                Bakery   
12                       The Danforth West, Riverdale      Greek Restaurant   
14       Brockton, Parkdale Village, Exhibition Place                  Café   
15                     India Bazaar, The Beaches West        Sandwich Place   
22                      High Park, The Junction South    Mexican Restaurant   
24                The Annex, North Midtown, Yorkville                  Café   
25                             Parkdale, Roncesvalles             Gift Shop   
26                                         Davisville           Pizza Place   
27                     University of Toronto, Harbord                  Café   
28                                 Runnymede, Swansea                  Café   
29                        Moore Park, Summerhill East                 Trail   
31  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
35                        St. James Town, Cabbagetown           Coffee Shop   
38  Business reply mail Processing Centre, South C...    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue   4th Most Common Venue  \
4                    Pub      Health Food Store                Wine Bar   
7                   Café                   Park      Athletics & Sports   
9               Pharmacy          Grocery Store             Supermarket   
12           Coffee Shop     Italian Restaurant  Furniture / Home Store   
14        Breakfast Spot  Performing Arts Venue             Coffee Shop   
15           Pizza Place   Fast Food Restaurant                     Gym   
22                  Café        Thai Restaurant             Flea Market   
24        Sandwich Place            Coffee Shop          History Museum   
25        Breakfast Spot     Italian Restaurant             Coffee Shop   
26        Sandwich Place           Dessert Shop        Sushi Restaurant   
27            Restaurant                    Bar     Japanese Restaurant   
28           Pizza Place            Coffee Shop                     Pub   
29            Playground            Summer Camp              Restaurant   
31                   Pub       Sushi Restaurant          Sandwich Place   
35            Restaurant                   Café                  Bakery   
38         Garden Center             Skate Park                 Brewery   

    5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
4            Cupcake Shop           Doner Restaurant               Dog Run   
7             Candy Store         Italian Restaurant            Baby Store   
9                    Bank                        Bar                  Café   
12             Restaurant             Ice Cream Shop        Cosmetics Shop   
14                 Bakery               Climbing Gym         Burrito Place   
15             Steakhouse             Ice Cream Shop      Sushi Restaurant   
22    Fried Chicken Joint                        Bar                Bakery   
24              BBQ Joint  Middle Eastern Restaurant     Indian Restaurant   
25          Movie Theater           Cuban Restaurant            Restaurant   
26            Coffee Shop         Italian Restaurant                  Café   
27              Bookstore             Sandwich Place                Bakery   
28     Italian Restaurant           Sushi Restaurant           Yoga Studio   
29               Wine Bar           Cuban Restaurant               Dog Run   
31  Vietnamese Restaurant         Light Rail Station   Fried Chicken Joint   
35                    Pub         Italian Restaurant           Pizza Place   
38                    Spa              Burrito Place            Restaurant   

        8th Most Common Venue      9th Most Common Venue  \
4         Distribution Center     

In [49]:
tor_merged.loc[tor_merged['Cluster Labels'] == 4, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
18  Lawrence Park           Swim School                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18    Dim Sum Restaurant              Bus Line              Wine Bar   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
18      Doner Restaurant               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
18        Discount Store                  Diner

## Analysis Conclusion
    I made a conclusion based on most common venue amoung neighbourhoods with eye-sight. 
    If you can give you inputs on how to make a proper conclusion based on data is much appreciated. 
    
    Cluster 1 : Coffe shops & Restaurants
    Cluster 2 : Park & Trails
    Cluster 3 : Home shopping 
    Cluster 4 : Restaurants & Cafe's
    Cluster 5 : SCommunity